File: go/Assembler/template_engine/engine_preprocess.go
package template_engine

import (
	"assembler/template_common"
	"assembler/template_model"
	"strings"
)

type EnginePreProcess struct {
	AppViewPrefix string
}

func NewEnginePreProcess(prefix string) *EnginePreProcess {
	return &EnginePreProcess{AppViewPrefix: prefix}
}

func (e *EnginePreProcess) SetAppViewPrefix(prefix string) {
	e.AppViewPrefix = prefix
}

func (e *EnginePreProcess) GetAppViewPrefix() string {
	return e.AppViewPrefix
}

// MergeTemplates merges templates using preprocessed data structures
func (e *EnginePreProcess) MergeTemplates(appSite, appFile, appView string, preprocessedTemplates map[string]template_model.PreprocessedTemplate, enableJsonProcessing bool) string {
	if len(preprocessedTemplates) == 0 {
		return ""
	}
	mainPreprocessed := e.GetTemplate(appSite, appFile, preprocessedTemplates, appView, e.AppViewPrefix, true)
	if mainPreprocessed == nil {
		return ""
	}
	contentHtml := mainPreprocessed.OriginalContent

	// Apply ALL replacement mappings from ALL templates (TemplateLoader did all the processing)
	contentHtml = e.ApplyTemplateReplacements(contentHtml, preprocessedTemplates, enableJsonProcessing, appView)

	return contentHtml
}

// GetTemplate retrieves a preprocessed template from the dictionary with AppView fallback
func (e *EnginePreProcess) GetTemplate(appSite, templateName string, preprocessedTemplates map[string]template_model.PreprocessedTemplate, appView, appViewPrefix string, useAppViewFallback bool) *template_model.PreprocessedTemplate {
	if len(preprocessedTemplates) == 0 {
		return nil
	}
	viewPrefix := appViewPrefix
	if viewPrefix == "" {
		viewPrefix = e.AppViewPrefix
	}

	// FIRST: Check for AppView-specific template resolution when AppView context is provided
	if useAppViewFallback && appView != "" && viewPrefix != "" {
		// Case-insensitive check if template_name contains view_prefix
		templateNameLower := strings.ToLower(templateName)
		viewPrefixLower := strings.ToLower(viewPrefix)

		if strings.Contains(templateNameLower, viewPrefixLower) {
			// Direct replacement: Replace the AppViewPrefix with the AppView value
			// For example: Html3AContent with AppViewPrefix=Html3A and AppView=html3B becomes html3BContent
			appKey := template_common.ReplaceCaseInsensitive(templateName, viewPrefix, appView)
			fallbackTemplateKey := strings.ToLower(appSite) + "_" + strings.ToLower(appKey)
			if tmpl, ok := preprocessedTemplates[fallbackTemplateKey]; ok {
				return &tmpl // Found AppView-specific template, use it
			}
		}
	}

	// SECOND: If no AppView-specific template found, try primary template
	primaryTemplateKey := strings.ToLower(appSite) + "_" + strings.ToLower(templateName)
	if tmpl, ok := preprocessedTemplates[primaryTemplateKey]; ok {
		return &tmpl
	}

	return nil
}

// ApplyTemplateReplacements applies all replacement mappings from all templates
func (e *EnginePreProcess) ApplyTemplateReplacements(content string, preprocessedTemplates map[string]template_model.PreprocessedTemplate, enableJsonProcessing bool, appView string) string {
	result := content
	maxPasses := 10

	for i := 0; i < maxPasses; i++ {
		previous := result

		for _, tmpl := range preprocessedTemplates {
			// Slotted templates first
			for _, mapping := range tmpl.ReplacementMappings {
				if mapping.Type == template_model.SlottedTemplateType && strings.Contains(result, mapping.OriginalText) {
					result = strings.ReplaceAll(result, mapping.OriginalText, mapping.ReplacementText)
				}
			}
			// Simple templates with AppView logic
			for _, mapping := range tmpl.ReplacementMappings {
				if mapping.Type == template_model.SimpleTemplateType && strings.Contains(result, mapping.OriginalText) {
					replacementText := e.ApplyAppViewLogicToReplacement(mapping.OriginalText, mapping.ReplacementText, preprocessedTemplates, appView)
					result = strings.ReplaceAll(result, mapping.OriginalText, replacementText)
				}
			}
			// JSON placeholders
			if enableJsonProcessing {
				for _, mapping := range tmpl.ReplacementMappings {
					if mapping.Type == template_model.JsonPlaceholderType && strings.Contains(result, mapping.OriginalText) {
						result = strings.ReplaceAll(result, mapping.OriginalText, mapping.ReplacementText)
					}
				}
				for _, placeholder := range tmpl.JsonPlaceholders {
					result = replaceAllCaseInsensitivePreProcess(result, placeholder.Placeholder, placeholder.Value)
				}
			}
		}

		if result == previous {
			break
		}
	}

	return result
}

// replaceCaseInsensitive replaces the first occurrence of 'from' in 'text' (case-insensitive) with 'to'
func replaceCaseInsensitive(text, from, to string) string {
	textLower := strings.ToLower(text)
	fromLower := strings.ToLower(from)
	if idx := strings.Index(textLower, fromLower); idx != -1 {
		end := idx + len(from)
		return text[:idx] + to + text[end:]
	}
	return text
}

// replaceAllCaseInsensitivePreProcess replaces all occurrences of a substring with another, ignoring case
func replaceAllCaseInsensitivePreProcess(input, search, replacement string) string {
	if search == "" {
		return input
	}
	lowerSearch := strings.ToLower(search)
	var builder strings.Builder
	lastIndex := 0
	for {
		index := strings.Index(strings.ToLower(input[lastIndex:]), lowerSearch)
		if index == -1 {
			builder.WriteString(input[lastIndex:])
			break
		}
		index += lastIndex
		builder.WriteString(input[lastIndex:index])
		builder.WriteString(replacement)
		lastIndex = index + len(search)
	}
	return builder.String()
}

// ApplyAppViewLogicToReplacement applies AppView fallback logic to template replacement text
func (e *EnginePreProcess) ApplyAppViewLogicToReplacement(originalText, replacementText string, preprocessedTemplates map[string]template_model.PreprocessedTemplate, appView string) string {
	placeholderName := extractPlaceholderName(originalText)
	if placeholderName == "" {
		return replacementText
	}

	var appSite string
	for key := range preprocessedTemplates {
		parts := strings.Split(key, "_")
		if len(parts) > 0 {
			appSite = parts[0]
			break
		}
	}

	template := e.GetTemplate(appSite, placeholderName, preprocessedTemplates, appView, e.AppViewPrefix, true)
	if template != nil {
		return template.OriginalContent
	}

	return replacementText
}

// extractPlaceholderName extracts placeholder name from {{PlaceholderName}} format
func extractPlaceholderName(originalText string) string {
	if !strings.HasPrefix(originalText, "{{") || !strings.HasSuffix(originalText, "}}") {
		return ""
	}
	return strings.TrimSpace(originalText[2 : len(originalText)-2])
}

File: rust/Assembler/src/template_engine/engine_preprocess.rs
use crate::template_model::model_preprocess::{PreprocessedTemplate, ReplacementType};
use crate::template_common::template_utils::TemplateUtils;
use std::collections::HashMap;

/// <summary>
/// PreProcess template engine implementation that only does merging using preprocessed data structures
/// All parsing is done by TemplateLoader, this engine only handles merging
/// </summary>
pub struct EnginePreProcess {
    app_view_prefix: String,
}

impl EnginePreProcess {
    pub fn new(prefix: String) -> Self {
        Self { app_view_prefix: prefix }
    }

    pub fn app_view_prefix(&self) -> &str {
        &self.app_view_prefix
    }
    
    pub fn set_app_view_prefix(&mut self, prefix: String) {
        self.app_view_prefix = prefix;
    }
    
    /// <summary>
    /// Merges templates using preprocessed data structures
    /// This method only does merging using preprocessed data structures - no loading or parsing
    /// </summary>
    /// <param name="app_site">The application site name for template key generation</param>
    /// <param name="app_file">The application file name</param>
    /// <param name="app_view">The application view name (optional)</param>
    /// <param name="preprocessed_templates">Dictionary of preprocessed templates for this specific appSite</param>
    /// <param name="enable_json_processing">Whether to enable JSON data processing</param>
    /// <returns>HTML with placeholders replaced using preprocessed structures</returns>
    pub fn merge_templates(
        &self,
        app_site: &str,
        app_file: &str,
        app_view: Option<&str>,
        preprocessed_templates: &HashMap<String, PreprocessedTemplate>,
        enable_json_processing: bool,
    ) -> String {
        if preprocessed_templates.is_empty() {
            return String::new();
        }

        // Use the get_template method to retrieve the main template
        let main_preprocessed = self.get_template(
            app_site,
            app_file,
            preprocessed_templates,
            app_view,
            Some(&self.app_view_prefix),
            true,
        );
        
        if main_preprocessed.is_none() {
            return String::new();
        }
        
        let main_preprocessed = main_preprocessed.unwrap();

        // Start with original content
        let content_html = main_preprocessed.original_content.clone();

        // Apply ALL replacement mappings from ALL templates (TemplateLoader did all the processing)
        self.apply_template_replacements(&content_html, preprocessed_templates, enable_json_processing, app_view)
    }

    /// <summary>
    /// Retrieves a template from the preprocessed templates dictionary based on various scenarios including AppView fallback logic
    /// </summary>
    /// <param name="app_site">The application site name</param>
    /// <param name="template_name">The template name (can be appFile or placeholderName)</param>
    /// <param name="preprocessed_templates">Dictionary of preprocessed templates</param>
    /// <param name="app_view">The application view name (optional)</param>
    /// <param name="app_view_prefix">The application view prefix (optional, uses instance property if not provided)</param>
    /// <param name="use_app_view_fallback">Whether to apply AppView fallback logic</param>
    /// <returns>The template's original content if found, None otherwise</returns>
    fn get_template< 'a>(
        &self,
        app_site: &str,
        template_name: &str,
        preprocessed_templates: &'a HashMap<String, PreprocessedTemplate>,
        app_view: Option<&str>,
        app_view_prefix: Option<&str>,
        use_app_view_fallback: bool,
    ) -> Option<&'a PreprocessedTemplate> {
        if preprocessed_templates.is_empty() {
            return None;
        }

        let view_prefix = app_view_prefix.unwrap_or(&self.app_view_prefix);
        
        // FIRST: Check for AppView-specific template resolution when AppView context is provided
        if use_app_view_fallback {
            if let (Some(app_view), Some(view_prefix)) = (app_view, Some(view_prefix)) {
                if !view_prefix.is_empty() && template_name.to_lowercase().contains(&view_prefix.to_lowercase()) {
                    let app_key = TemplateUtils::replace_case_insensitive(template_name, view_prefix, app_view);
                    let fallback_template_key = format!("{}_{}", app_site.to_lowercase(), app_key.to_lowercase());
                    if let Some(fallback_template) = preprocessed_templates.get(&fallback_template_key) {
                        return Some(fallback_template);
                    }
                }
            }
        }

        // SECOND: If no AppView-specific template found, try primary template
        let primary_template_key = format!("{}_{}", app_site.to_lowercase(), template_name.to_lowercase());
        if let Some(primary_template) = preprocessed_templates.get(&primary_template_key) {
            return Some(primary_template);
        }

        None
    }

    /// <summary>
    /// Applies all replacement mappings from all templates - NO processing logic, only direct replacements
    /// </summary>
    fn apply_template_replacements(
        &self,
        content: &str,
        preprocessed_templates: &HashMap<String, PreprocessedTemplate>,
        enable_json_processing: bool,
        app_view: Option<&str>,
    ) -> String {
        let mut result = content.to_string();

        let mut previous;
        let max_passes = 10;
        let mut current_pass = 0;
        
        loop {
            previous = result.clone();
            current_pass += 1;
            
            for template in preprocessed_templates.values() {
                for mapping in template.replacement_mappings.iter().filter(|m| matches!(m.r#type, ReplacementType::SlottedTemplate)) {
                    if result.contains(&mapping.original_text) {
                        result = result.replace(&mapping.original_text, &mapping.replacement_text);
                    }
                }
                
                for mapping in template.replacement_mappings.iter().filter(|m| matches!(m.r#type, ReplacementType::SimpleTemplate)) {
                    if result.contains(&mapping.original_text) {
                        let replacement_text = self.apply_app_view_logic_to_replacement(&mapping.original_text, &mapping.replacement_text, preprocessed_templates, app_view);
                        result = result.replace(&mapping.original_text, &replacement_text);
                    }
                }
                
                if enable_json_processing {
                    for mapping in template.replacement_mappings.iter().filter(|m| matches!(m.r#type, ReplacementType::JsonPlaceholder)) {
                        if result.contains(&mapping.original_text) {
                            result = result.replace(&mapping.original_text, &mapping.replacement_text);
                        }
                    }
                }
                
                if enable_json_processing {
                    for placeholder in &template.json_placeholders {
                        result = Self::replace_all_case_insensitive(&result, &placeholder.placeholder, &placeholder.value);
                    }
                }
            }
            
            if result == previous || current_pass >= max_passes {
                break;
            }
        }

        result
    }

    /// <summary>
    /// Helper method to replace all case-insensitive occurrences
    /// </summary>
    fn replace_all_case_insensitive(input: &str, search: &str, replacement: &str) -> String {
        let mut result = input.to_string();
        let mut idx = 0;
        while let Some(found) = result[idx..].to_lowercase().find(&search.to_lowercase()) {
            let found = idx + found;
            result.replace_range(found..found + search.len(), replacement);
            idx = found + replacement.len();
        }
        result
    }

    /// <summary>
    /// Applies AppView fallback logic to template replacement text using the centralized get_template method
    /// </summary>
    fn apply_app_view_logic_to_replacement(
        &self,
        original_text: &str,
        replacement_text: &str,
        preprocessed_templates: &HashMap<String, PreprocessedTemplate>,
        app_view: Option<&str>,
    ) -> String {
        let placeholder_name = Self::extract_placeholder_name(original_text);
        if placeholder_name.is_empty() {
            return replacement_text.to_string();
        }

        let sample_key = match preprocessed_templates.keys().next() {
            Some(key) => key,
            None => return replacement_text.to_string(),
        };
        
        let parts: Vec<&str> = sample_key.split('_').collect();
        if parts.len() < 2 {
            return replacement_text.to_string();
        }
        
        let app_site = parts[0];
        
        let template = self.get_template(app_site, &placeholder_name, preprocessed_templates, app_view, Some(&self.app_view_prefix), true);
        
        template.map(|t| t.original_content.clone()).unwrap_or_else(|| replacement_text.to_string())
    }

    /// <summary>
    /// Extracts placeholder name from {{PlaceholderName}} format
    /// </summary>
    fn extract_placeholder_name(original_text: &str) -> String {
        if !original_text.starts_with("{{") || !original_text.ends_with("}}") {
            return String::new();
        }
        
        original_text[2..original_text.len() - 2].trim().to_string()
    }
}
